File: NSNotifsExample/Objective-C/Controllers/FirstVC.m
#import "FirstVC.h"


@implementation FirstVC {

	UIButton *magicButton;

}


- (id)initWithNibName:(NSString *)aNib bundle:(NSBundle *)aBundle {

	self = [super initWithNibName:aNib bundle:aBundle];

	if(self) {

		// Custom initialization

		[self setupUI];

	}

	return self;

}


- (void)viewDidLoad {

	[super viewDidLoad];

	// Do any additional setup after loading the view, typically from a nib.

	self.view.backgroundColor = UIColor.systemBackgroundColor;

}


- (void)viewDidLayoutSubviews {

	[super viewDidLayoutSubviews];

	[self layoutUI];

}

- (void)setupUI {

	magicButton = [UIButton new];
	magicButton.titleLabel.font = [UIFont systemFontOfSize:15];
	magicButton.backgroundColor = UIColor.systemIndigoColor;
	magicButton.layer.cornerCurve = kCACornerCurveContinuous;
	magicButton.layer.cornerRadius = 20;
	magicButton.translatesAutoresizingMaskIntoConstraints = NO;
	[magicButton setTitle : @"Tap me and switch tabs to see magic" forState:UIControlStateNormal];
	[magicButton addTarget : self action:@selector(notificationSender:) forControlEvents:UIControlEventTouchUpInside];
	[self.view addSubview:magicButton];

}


- (void)layoutUI {

	NSDictionary *views = @{

		@"magicButton": magicButton,
		@"superview": self.view

	};

	NSString *formatCenterX = @"V:[magicButton]-(<=1)-[superview]";
	NSString *formatCenterY = @"H:[magicButton]-(<=1)-[superview]";
	NSString *formatWidth = @"H:[magicButton(==260)]";
	NSString *formatHeight = @"V:[magicButton(==40)]";

	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatCenterX options:NSLayoutFormatAlignAllCenterX metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatCenterY options:NSLayoutFormatAlignAllCenterY metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatWidth options:0 metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatHeight options:0 metrics:nil views:views]];

}


- (void)notificationSender:(id)sender {

	[NSNotificationCenter.defaultCenter postNotificationName:@"fireNotificationDone" object:nil];

}


@end

File: NSNotifsExample/Swift/Controllers/FirstVC.swift
import UIKit


final class FirstVC: UIViewController {

	/*--- instantiate UIButton so we can access all 
	of it's properties and create our button ---*/

	private let magicButton: UIButton = {
		let button = UIButton()
		button.setTitle("Tap me and switch tabs to see magic", for: .normal)
		button.backgroundColor = .systemIndigo
		button.titleLabel?.font = UIFont.systemFont(ofSize: 15)
		button.layer.cornerCurve = .continuous
		button.layer.cornerRadius = 20
		button.translatesAutoresizingMaskIntoConstraints = false
		button.addTarget(self, action:#selector(notificationSender), for: .touchUpInside)
		return button
	}()

 	required init?(coder aDecoder: NSCoder) {

		super.init(coder: aDecoder)

	}

	init() {

		super.init(nibName: nil, bundle: nil)

		setupUI()

	}

	override func viewDidLoad() {

		super.viewDidLoad()

		// Do any additional setup after loading the view, typically from a nib.

		view.backgroundColor = .systemBackground

	}

	override func viewDidLayoutSubviews() {

		super.viewDidLayoutSubviews()

		layoutUI()

	}

	private func setupUI() {

		view.addSubview(magicButton)

	}

	private func layoutUI() {

		// proper UI layout

		let centerXConstraint = NSLayoutConstraint(item: magicButton, attribute: .centerX, relatedBy: .equal, toItem: 
		view, attribute: .centerX, multiplier: 1, constant: 0)

		let centerYConstraint = NSLayoutConstraint(item: magicButton, attribute: .centerY, relatedBy: .equal, toItem:
		view, attribute: .centerY, multiplier: 1, constant: 0)

		let widthConstraint = NSLayoutConstraint(item: magicButton, attribute: .width, relatedBy: .equal, toItem: 
		nil, attribute: .notAnAttribute, multiplier: 1, constant: 260)

		let heightConstraint = NSLayoutConstraint(item: magicButton, attribute: .height, relatedBy: .equal, toItem:
		nil, attribute: .notAnAttribute, multiplier: 1, constant: 40)

		view.addConstraints([centerXConstraint, centerYConstraint, widthConstraint, heightConstraint])

	}

	/*--- we use @objc because "selector" is an Objective-C concept only,
	so we need to make it visible to the Objective-C runtime ---*/

	@objc private func notificationSender() {

		// fire the notification when clicking the button

		NotificationCenter.default.post(name: Notification.Name("fireNotificationDone"), object: nil)
		
	}

}
